Sources/SKTestSupport/ProcessRunner.swift

```diff
@@ -0,0 +1,189 @@
+    }
+
+    var stderrData = Data()
+    if captureStderr {
+      let errPipe = Pipe()
+      process.standardError = errPipe
+      addHandler(pipe: errPipe, group: group) { stderrData.append($0) }
+    }
+
+    try process.run()
+    SigIntListener.runningSubprocesses.insert(process)
+    process.waitUntilExit()
+    SigIntListener.runningSubprocesses.remove(process)
+    if captureStdout || captureStderr {
+      // Make sure we've received all stdout/stderr
+      group.wait()
+    }
+
+    guard let stdoutString = String(data: stdoutData, encoding: .utf8) else {
+      throw FailedToDecodeUTF8Error(data: stdoutData)
+    }
+    guard let stderrString = String(data: stderrData, encoding: .utf8) else {
+      throw FailedToDecodeUTF8Error(data: stderrData)
+    }
+
+    guard process.terminationStatus == 0 else {
+      throw NonZeroExitCodeError(
+        process: process,
+        stdout: stdoutString,
+        stderr: stderrString,
+        exitCode: Int(process.terminationStatus)
+      )
+    }
+
+    return ProcessResult(
+      stdout: stdoutString,
+      stderr: stderrString
+    )
+  }
+
+  private func addHandler(
+    pipe: Pipe,
+    group: DispatchGroup,
+    addData: @escaping (Data) -> Void
+  ) {
+    group.enter()
+    pipe.fileHandleForReading.readabilityHandler = { fileHandle in
+      // Apparently using availableData can cause various issues
+      let newData = fileHandle.readData(ofLength: Int.max)
+      if newData.count == 0 {
+        pipe.fileHandleForReading.readabilityHandler = nil;
+        group.leave()
+      } else {
+        addData(newData)
+      }
+    }
+  }
+}
+
+/// The exit code and output (if redirected) from a subprocess that has
+/// terminated
+public struct ProcessResult {
+  public let stdout: String
+  public let stderr: String
+}
+
+/// Error thrown if a process terminates with a non-zero exit code.
+struct NonZeroExitCodeError: Error, CustomStringConvertible {
+  let process: Process
+  let stdout: String
+  let stderr: String
+  let exitCode: Int
+
+  var description: String {
+    var result = """
+      Command failed with non-zero exit code \(exitCode):
+      Command: \(process.command)
+      """
+    if !stdout.isEmpty {
+      result += """
+        Standard output:
+        \(stdout)
+        """
+    }
+    if !stderr.isEmpty {
+      result += """
+        Standard error:
+        \(stderr)
+        """
+    }
+    return result
+  }
+}
+
+/// Error thrown if `stdout` or `stderr` could not be decoded as UTF-8.
+struct FailedToDecodeUTF8Error: Error {
+  let data: Data
+}
+
+extension Process {
+  var command: String {
+    var message = ""
+
+    for (key, value) in environment?.sorted(by: { $0.key < $1.key }) ?? [] {
+      message += "\(key)='\(value)' "
+    }
+
+    if let executableURL = executableURL {
+      message += executableURL.path
+    }
+
+    if let arguments = arguments {
+      message += " \(arguments.joined(separator: " "))"
+    }
+
+    return message
+  }
```

Comment:
Do we need to worry about escaping here, or does it not matter since it looks like this is currently only used for verbose printing? Would it be worth naming it `commandDescription` or something to make it clear that it's only meant for printing?

Reply:
Yes, it’s only used for verbose printing an it’s been good enough so far. I’ll add a comment that it’s only meant for printing.